Support client-side dynamic routes in static export

matheusslg opened the pull request:
## Summary

Adds support for client-side dynamic routes in Next.js static export mode, addressing issue #64660.

Pages marked with `'use client'` can now use dynamic route segments without pre-generating every possible URL combination at build time. The implementation generates shell templates with placeholders that are patched with actual parameter values at runtime in the browser.

## Use Cases

- Dynamic routes with too many parameter combinations to pre-generate
- Route parameters from user input or external data sources
- SPA-style routing in fully static Next.js exports

## How It Works

### Build Time

1. **Shell Files**: Generates template files with `__shell__` placeholders for dynamic segments
   ```
   out/blog/__shell__.html              # For [slug]
   out/products/__shell__/__shell__.html # For [category]/[id]
   out/docs/__shell__.html              # For [[...slug]]
   ```

2. **Manifest**: Creates `_clientDynamicRoutesManifest.json` mapping route patterns to their segment structure

3. **SPA Fallback**: Generates `404.html` with external script reference for CSP compliance

### Runtime

When accessing a dynamic route URL directly:

1. Static server returns `404.html`
2. External script loads manifest and matches URL pattern
3. Fetches appropriate shell file
4. Patches RSC data with actual parameter values using streaming
5. React renders the page with correct parameters

Client-side navigation works seamlessly without page reloads.

## Supported Patterns

- Single parameter: `/blog/[slug]`
- Multiple parameters: `/products/[category]/[id]`
- Required catch-all: `/files/[...path]` (1+ segments required)
- Optional catch-all: `/docs/[[...slug]]` (0+ segments)

## Example

```tsx
'use client'

import { useParams } from 'next/navigation'

export default function BlogPost() {
  const params = useParams()
  return <article>Post: {params.slug}</article>
}
```

With `output: 'export'`, visiting `/blog/my-post` directly will:
1. Load `404.html` 
2. Fetch `/blog/__shell__.html`
3. Patch RSC data: `"slug":"__shell__"` → `"slug":"my-post"`
4. Render the page with correct parameter

## Testing

Added comprehensive test suite (`test/integration/app-dir-export/test/client-dynamic-route.test.ts`):

- Shell file generation with placeholders
- External CSP-compliant fallback script
- Manifest structure with optional field support
- Client-side navigation between routes
- Content integrity for literal `__shell__` text
- Multiple parameter combinations

All 8 tests passing ✅

## Documentation

Updated `docs/01-app/02-guides/static-exports.mdx` with:
- Complete explanation of client-side dynamic routes
- Build output structure
- How the SPA fallback mechanism works
- All supported route patterns
- Important considerations (CSP, error handling, React hydration)

## Closes

Closes #64660

matheusslg committed:
Support client-side dynamic routes in static export

Implements client-side dynamic route handling for Next.js static export, addressing issue #64660.

This adds support for dynamic routes that are rendered client-side at runtime rather than being pre-rendered during build. This is useful for:
- Dynamic routes with too many possible parameter values to pre-generate
- Routes where parameters come from user input or external data
- Scenarios where full static generation isn't practical

Implementation:
1. New opt-in marker `export const dynamicParams = 'client'` for dynamic routes
2. Generates a single "__shell__" file containing placeholder params
3. Creates a manifest mapping dynamic route patterns to their segments
4. Custom 404.html that loads shell and patches params from URL at runtime
5. Smart prefetch that fetches shell instead of specific params (no 404s)
6. Client-side param patching in fetch-server-response for SPA navigation

Benefits:
- Avoids generating thousands of static files for dynamic routes
- Maintains SPA-style navigation without full page reloads
- No 404 errors during prefetch
- Single shell file serves all parameter variations

Tests:
- Added comprehensive test suite covering:
  - Shell file generation
  - 404 fallback with param patching
  - Client-side navigation
  - Multiple dynamic parameters

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 4fb346b918e4c86c562249efe98e449b81578963

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

matheusslg committed:
Merge branch 'canary' into feat/client-dynamic-routes-static-export

matheusslg committed:
Fix React hydration for client dynamic routes in static export

This commit fixes the critical hydration issue where React App Router
ignores patched RSC data on initial page load.

Changes:
1. Generate minimal shell HTML for client dynamic routes
   - Detect shell files by checking for __shell__ in export path
   - Keep head and RSC scripts but replace body with minimal content
   - Forces React to do client-side render instead of hydration

2. Fix 404.html to use document.write() for page replacement
   - Replace entire page before React initializes
   - Ensures React sees patched RSC data from the start
   - Eliminates timing issues with DOM manipulation

How it works:
- Build time: Generate shell files with minimal HTML
- Direct URL access: Browser gets 404.html
- 404.html script: Fetches shell, patches RSC data, replaces page
- React initializes: Sees patched RSC data and renders correctly

matheusslg committed:
Add catch-all route support for client dynamic routes

This commit extends client-side navigation to support catch-all
routes ([...slug]) in static export mode.

Changes:
- Support catch-all segments in route matching
- Handle array param values (e.g., slug: ['a', 'b', 'c'])
- Improve param patching to use param-specific regex
- Join catch-all arrays to paths (a/b/c)

This allows routes like /blog/[...slug] to work correctly with
client-side navigation in static export mode.

matheusslg committed:
Improve test server file resolution for shell files

Enhance serve-404.js to try multiple file path variations:
- Direct path (e.g., /client-dynamic/__shell__)
- Path with .html extension
- Path with /index.html

This makes it easier to test shell file generation and
ensures the test server can serve all static export files.

matheusslg committed:
Fix critical issues in client dynamic routes static export

- Fix content corruption by removing global __shell__ replacement
  Use param-specific regex only to avoid corrupting content that
  contains the literal text "__shell__"

- Fix CSP violations by extracting inline JavaScript
  Move ~200 lines of inline JS from 404.html to external script
  (app-client-dynamic-fallback.js) to comply with Content Security Policy

- Implement stream-based RSC patching for memory efficiency
  Use TransformStream with 200-char lookbehind buffer to patch
  chunks on-the-fly instead of reading entire payload into memory

- Add manifest retry logic with exponential backoff
  Retry failed manifest loads up to 3 times (100ms, 500ms, 1000ms)
  to prevent permanent failures from temporary network issues

- Cache AST parsing to improve build performance
  Reuse parsed AST when checking for 'use client' directive
  to avoid parsing the same file twice

- Add timeout handling and error UX
  10-second timeout for all fetches with user-friendly error
  messages and retry button

- Add <noscript> fallback for accessibility
  Show message when JavaScript is disabled

- Add test case for content integrity
  Verify content containing "__shell__" text is not corrupted
  during RSC patching

matheusslg committed:
Update static export docs for client-side dynamic routes

Clarify how client-side dynamic routes work in static export mode:
- Shell files with __shell__ placeholders are pre-rendered at build
- 404.html acts as SPA fallback for direct URL access
- Manifest and external script handle route matching and patching
- Document supported patterns (single/multi params, catch-all)
- Add CSP compliance and error handling details
- Note React hydration warning #418 is expected behavior

matheusslg committed:
Fix optional catch-all routes in client dynamic routes static export

Add support for optional catch-all routes like [[...slug]] which were
previously failing to match when no segments were provided.

Changes:
- Add optional field to TypeScript interface for manifest segments
- Rewrite route matching logic to handle optional parameters
- Calculate minimum required segments (static + non-optional dynamic)
- Support optional catch-all with empty arrays []
- Support optional regular dynamic segments
- Update 404.html fallback script with same logic
- Update tests to reflect new implementation (external CSP-compliant script)
- Update documentation to explain how it works

Fixes issue where routes like /docs with [[...slug]] would return 404
instead of matching with { slug: [] }

matheusslg committed:
Fix regex escaping bug in dynamic route shell generation

Escape regex special characters in parameter names before using them
in RegExp patterns to prevent incorrect replacements.

Issue:
Parameter names containing regex metacharacters (*, +, $, (), etc.)
were not being replaced with __shell__ placeholders, leaving shell
pathnames malformed.

Example:
- Route: /app/[item*]/page
- Before: /app/[item*]/page (not replaced!)
- After: /app/__shell__/page (correctly replaced)

Solution:
Use String.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') to escape all
regex special characters in the parameter name before constructing
the RegExp pattern.

matheusslg committed:
Fix optional catch-all routes in client dynamic routes static export

Generate both root and shell files for optional catch-all routes.

Before this fix, routes like /docs/[[...slug]] only generated the
shell file (/docs/__shell__/index.html) but not the root case file
(/docs/index.html). This caused accessing /docs to show directory
listings or 404 errors.

Now optional catch-all routes generate two static path entries:
1. Shell file with __shell__ placeholder for parameterized access
2. Root file without the param for the empty/root case

This ensures both /docs and /docs/something work correctly.

BrodyEller commented:
Looking really good man, any way I can help?

sovetski commented:
@timneutkens @ijjk @sokra @huozhi we really need this PR, look at https://github.com/vercel/next.js/discussions/64660

matheusslg commented:
Hey everyone — apologies for the delay. I haven’t had much time to revisit this. From what I recall and from the progress I made, the main issue was that I wasn’t achieving true SPA navigation; the page was fully reloading instead of transitioning client-side, which isn’t ideal. The dynamic navigation with custom parameters was working as expected, but I was running into some React hydration warnings in the console that I haven’t investigated further yet.

If anyone would like to review the PR, explore the ideas, or even fork it for experimentation, feel free to do so.
